src/app/frontend/common/services/global/history.ts

```diff
@@ -0,0 +1,47 @@
+// Copyright 2017 The Kubernetes Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {Injectable} from '@angular/core';
+import {RawParams, StateService, TransitionPromise, TransitionService} from '@uirouter/core';
+
+@Injectable()
+export class HistoryService {
+  private previousStateName: string;
+  private previousStateParams: RawParams;
+
+  constructor(
+      private readonly state: StateService, private readonly transitions: TransitionService) {}
+
+  /** Initializes the service. Must be called before use. */
+  init(): void {
+    this.transitions.onSuccess({}, (transition) => {
```

Comment:
You can subscribe to one of the wrappers exposed in [KdStateService](https://github.com/pengx17/dashboard/blob/b5f84015da58e752832953883a92da4b2be0045f/src/app/frontend/common/services/global/state.ts#L21). It's done this way so we don't have to use this global initialization.

Reply:
OK. I guess we should still rely on ui-router's `StateService` to do navigations (except href) like `go`?

Reply:
In this case, yes. It is required to navigate through lazy loaded modules as a workaround for issue reported on uirouter's repo. In case where module was already loaded, which is the case for previous state, then uirouter will correctly handle transitions.

Reply:
@floreks I see `KdStateService` does not expose transition object (of ui-router). Is this by design?

Reply:
It exposes concrete transitions like `onBefore` and `onSuccess`. It is enough for 90% of cases. I don't think we will need more.